File: src/ingest/chunkers.ts
export interface Chunk {
  ordinal: number;
  headingPath: string | null;
  content: string;
}

export interface ChunkOptions {
  maxChars?: number;
  minChars?: number;
}

interface ResolvedOptions {
  maxChars: number;
  minChars: number;
}

interface RawChunk {
  headingPath: string | null;
  content: string;
}

const DEFAULT_MAX_CHARS = 1200;
const DEFAULT_MIN_CHARS = 200;

const ATX_HEADING = /^(#{1,6})\s+(.+?)\s*#*\s*$/;

function resolveOptions(opts: ChunkOptions): ResolvedOptions {
  const maxChars =
    typeof opts.maxChars === "number" &&
    Number.isFinite(opts.maxChars) &&
    opts.maxChars > 0
      ? Math.floor(opts.maxChars)
      : DEFAULT_MAX_CHARS;
  const minChars =
    typeof opts.minChars === "number" &&
    Number.isFinite(opts.minChars) &&
    opts.minChars > 0
      ? Math.floor(opts.minChars)
      : DEFAULT_MIN_CHARS;
  return { maxChars, minChars };
}

/**
 * Split text into pieces of at most maxChars characters.
 * Tier 1: paragraphs (blank-line separated). Tier 2: lines. Tier 3: hard slices.
 */
function splitToSize(text: string, maxChars: number): string[] {
  const trimmed = text.trim();
  if (trimmed.length === 0) return [];
  return splitTier(trimmed, maxChars);
}

function splitTier(text: string, maxChars: number): string[] {
  if (text.length <= maxChars) return [text];
  const parts = text
    .split(/\n{2,}/)
    .map((part) => part.trim())
    .filter((part) => part.length > 0);
  if (parts.length > 1)
    return packPieces(
      parts.flatMap((part) => splitTier(part, maxChars)),
      "\n\n",
      maxChars,
    );
  const lines = text.split("\n").map((line) => line.trimEnd());
  if (lines.length > 1)
    return packPieces(
      lines.flatMap((line) => splitLine(line, maxChars)),
      "\n",
      maxChars,
    );
  return splitLine(text, maxChars);
}

function splitLine(line: string, maxChars: number): string[] {
  if (line.length <= maxChars) return [line];
  const slices: string[] = [];
  for (let i = 0; i < line.length; i += maxChars) {
    slices.push(line.slice(i, i + maxChars));
  }
  return slices;
}

function packPieces(
  pieces: string[],
  separator: string,
  maxChars: number,
): string[] {
  const out: string[] = [];
  let buffer = "";
  for (const piece of pieces) {
    if (buffer.length === 0) {
      buffer = piece;
    } else if (buffer.length + separator.length + piece.length <= maxChars) {
      buffer += separator + piece;
    } else {
      out.push(buffer);
      buffer = piece;
    }
  }
  if (buffer.length > 0) out.push(buffer);
  return out;
}

// ---------------------------------------------------------------------------
// Markdown chunking
// ---------------------------------------------------------------------------

interface SectionAccumulator {
  path: string[] | null;
  lines: string[];
}

function splitMarkdownSections(body: string): RawChunk[] {
  const sections: RawChunk[] = [];
  const stack: Array<{ level: number; title: string }> = [];
  let current: SectionAccumulator = { path: null, lines: [] };

  const flush = (): void => {
    const content = current.lines.join("\n").trim();
    if (content.length > 0) {
      sections.push({
        headingPath: current.path === null ? null : current.path.join(" > "),
        content,
      });
    }
  };

  for (const line of body.split("\n")) {
    const heading = line.match(ATX_HEADING);
    if (heading === null) {
      current.lines.push(line);
      continue;
    }
    flush();
    const level = heading[1].length;
    const title = heading[2].trim();
    while (stack.length > 0 && stack[stack.length - 1].level >= level)
      stack.pop();
    stack.push({ level, title });
    current = { path: stack.map((entry) => entry.title), lines: [] };
  }
  flush();
  return sections;
}

function mergeSmallChunks(
  chunks: RawChunk[],
  minChars: number,
  maxChars: number,
): RawChunk[] {
  const merged: RawChunk[] = [];
  for (const chunk of chunks) {
    const previous = merged[merged.length - 1];
    if (
      previous !== undefined &&
      chunk.content.length < minChars &&
      previous.headingPath === chunk.headingPath &&
      previous.content.length + chunk.content.length <= maxChars * 1.5
    ) {
      previous.content = `${previous.content}\n\n${chunk.content}`;
    } else {
      merged.push({ headingPath: chunk.headingPath, content: chunk.content });
    }
  }
  return merged;
}

export function chunkMarkdown(body: string, opts: ChunkOptions = {}): Chunk[] {
  const { maxChars, minChars } = resolveOptions(opts);
  const sections = splitMarkdownSections(body);
  const merged = mergeSmallChunks(sections, minChars, maxChars);
  const raw: RawChunk[] = [];
  for (const chunk of merged) {
    if (chunk.content.length <= maxChars) {
      raw.push(chunk);
    } else {
      for (const piece of splitToSize(chunk.content, maxChars)) {
        raw.push({ headingPath: chunk.headingPath, content: piece });
      }
    }
  }
  return raw.map((chunk, index) => ({
    ordinal: index,
    headingPath: chunk.headingPath,
    content: chunk.content,
  }));
}

// ---------------------------------------------------------------------------
// Code chunking
// ---------------------------------------------------------------------------

const LANGUAGE_BY_EXT: Record<string, string> = {
  ts: "typescript",
  tsx: "tsx",
  js: "javascript",
  jsx: "jsx",
  py: "python",
  go: "go",
  rs: "rust",
  java: "java",
  c: "c",
  h: "c",
  cc: "cpp",
  cpp: "cpp",
  hpp: "cpp",
  cs: "csharp",
  rb: "ruby",
  php: "php",
  sh: "shell",
  sql: "sql",
  md: "markdown",
  json: "json",
  yaml: "yaml",
  yml: "yaml",
  toml: "toml",
  html: "html",
  css: "css",
};

export function detectLanguage(filename: string): string | null {
  const base = filename.split(/[\\/]/).pop() ?? filename;
  const dot = base.lastIndexOf(".");
  if (dot <= 0) return null;
  return LANGUAGE_BY_EXT[base.slice(dot + 1).toLowerCase()] ?? null;
}

interface DeclMatch {
  indent: number;
  name: string;
  callable: boolean;
  lineIndex: number;
}

interface DeclNode extends DeclMatch {
  children: DeclNode[];
  spanEnd: number;
}

// Keyword-style declarations: export? abstract? async? pub? class/function/struct/impl/fn/func/def Name
const KEYWORD_DECL =
  /^[ \t]*(?:export\s+)?(?:default\s+)?(?:declare\s+)?(?:abstract\s+)?(?:async\s+)?(?:pub\s+)?(class|function|struct|impl|fn|func|def)\s+([A-Za-z_$][A-Za-z0-9_$]*)/;

// Visibility-modifier methods (C-family/TS): private [static] [async] name(
const MODIFIER_DECL =
  /^[ \t]*(?:public|private|protected)(?:\s+(?:static|readonly|async|override|final|abstract))*\s+([A-Za-z_$][A-Za-z0-9_$]*)\s*(?:<[^>\n]*>)?\(/;
// Plain indented methods (C-family/TS): name(params): RetType {
const METHOD_DECL =
  /^[ \t]+([A-Za-z_$][A-Za-z0-9_$]*)\s*(?:<[^>\n]*>)?\([^;{}]*\)\s*(?::\s*[^{;=]+)?\{/;

const RESERVED_NAMES: Record<string, true> = {
  if: true,
  for: true,
  while: true,
  switch: true,
  catch: true,
  return: true,
  new: true,
  function: true,
};

const CALLABLE_KEYWORDS: Record<string, true> = {
  function: true,
  fn: true,
  func: true,
  def: true,
};

const COMMENT_LINE = /^[ \t]*(\/\/|\/\*|\*|#)/;

function scanDecls(lines: string[], languageHint: string | null): DeclMatch[] {
  const pythonOnly = languageHint === "python";
  const decls: DeclMatch[] = [];
  for (let i = 0; i < lines.length; i++) {
    const line = lines[i];
    if (line === undefined || COMMENT_LINE.test(line)) continue;

    const keywordMatch = pythonOnly ? null : line.match(KEYWORD_DECL);
    if (keywordMatch !== null) {
      const keyword = keywordMatch[1];
      const name = keywordMatch[2];
      if (keyword === undefined || name === undefined) continue;
      decls.push({
        indent: line.length - line.trimStart().length,
        name,
        callable: CALLABLE_KEYWORDS[keyword] === true,
        lineIndex: i,
      });
      continue;
    }

    const pyMatch = line.match(
      /^[ \t]*(?:async\s+)?(def|class)\s+([A-Za-z_][A-Za-z0-9_]*)/,
    );
    if (pyMatch !== null) {
      const keyword = pyMatch[1];
      const name = pyMatch[2];
      if (keyword === undefined || name === undefined) continue;
      decls.push({
        indent: line.length - line.trimStart().length,
        name,
        callable: keyword === "def",
        lineIndex: i,
      });
      continue;
    }
    if (pythonOnly) continue;
    const modifierMatch = line.match(MODIFIER_DECL);
    if (modifierMatch !== null) {
      const name = modifierMatch[1];
      if (name === undefined) continue;
      decls.push({
        indent: line.length - line.trimStart().length,
        name,
        callable: true,
        lineIndex: i,
      });
      continue;
    }

    const methodMatch = line.match(METHOD_DECL);
    if (
      methodMatch !== null &&
      !line.includes("=>") &&
      methodMatch[1] !== undefined &&
      RESERVED_NAMES[methodMatch[1]] !== true
    ) {
      decls.push({
        indent: line.length - line.trimStart().length,
        name: methodMatch[1],
        callable: true,
        lineIndex: i,
      });
    }
  }
  return decls;
}

function buildTree(decls: DeclMatch[], totalLines: number): DeclNode[] {
  const roots: DeclNode[] = [];
  const stack: DeclNode[] = [];
  for (let i = 0; i < decls.length; i++) {
    const decl = decls[i];
    if (decl === undefined) continue;
    let spanEnd = totalLines;
    for (let j = i + 1; j < decls.length; j++) {
      const later = decls[j];
      if (later !== undefined && later.indent <= decl.indent) {
        spanEnd = later.lineIndex;
        break;
      }
    }
    const node: DeclNode = { ...decl, spanEnd, children: [] };
    while (stack.length > 0 && stack[stack.length - 1].indent >= node.indent)
      stack.pop();
    const parent = stack[stack.length - 1];
    if (parent === undefined) roots.push(node);
    else parent.children.push(node);
    stack.push(node);
  }
  return roots;
}

function displayName(node: DeclMatch): string {
  return node.callable ? `${node.name}()` : node.name;
}

function pushChunk(
  content: string,
  headingPath: string | null,
  out: RawChunk[],
  maxChars: number,
): void {
  if (content.length <= maxChars) {
    const trimmed = content.trim();
    if (trimmed.length > 0) out.push({ headingPath, content: trimmed });
    return;
  }
  for (const piece of splitToSize(content, maxChars)) {
    out.push({ headingPath, content: piece });
  }
}

function emitNode(
  lines: string[],
  node: DeclNode,
  ancestors: string[],
  out: RawChunk[],
  maxChars: number,
): void {
  const path = [...ancestors, displayName(node)];
  const headingPath = path.join(" > ");

  if (node.children.length > 0) {
    let cursor = node.lineIndex;
    for (const child of node.children) {
      const head = lines.slice(cursor, child.lineIndex).join("\n");
      pushChunk(head, headingPath, out, maxChars);
      emitNode(lines, child, path, out, maxChars);
      cursor = child.spanEnd;
    }
    const tail = lines.slice(cursor, node.spanEnd).join("\n");
    pushChunk(tail, headingPath, out, maxChars);
    return;
  }

  const body = lines.slice(node.lineIndex, node.spanEnd).join("\n");
  pushChunk(body, headingPath, out, maxChars);
}

export function chunkCode(
  content: string,
  languageHint: string | null,
  opts: ChunkOptions = {},
): Chunk[] {
  const { maxChars } = resolveOptions(opts);
  const lines = content.split("\n");
  const decls = scanDecls(lines, languageHint);

  if (decls.length === 0) {
    const raw: RawChunk[] = [];
    pushChunk(content, null, raw, maxChars);
    return raw.map((chunk, index) => ({
      ordinal: index,
      headingPath: chunk.headingPath,
      content: chunk.content,
    }));
  }

  const roots = buildTree(decls, lines.length);
  const raw: RawChunk[] = [];
  const first = decls[0];
  if (first !== undefined) {
    pushChunk(lines.slice(0, first.lineIndex).join("\n"), null, raw, maxChars);
  }
  for (const root of roots) emitNode(lines, root, [], raw, maxChars);
  return raw.map((chunk, index) => ({
    ordinal: index,
    headingPath: chunk.headingPath,
    content: chunk.content,
  }));
}

File: rust/src/ingest/chunkers.rs
//! Markdown and source-code chunkers for indexing.
//!
//! Markdown splits on ATX headings building full ancestor heading paths;
//! small consecutive chunks merge; oversized sections fall back through
//! paragraphs → lines → hard slices. Code chunking scans declaration lines
//! (brace-language keywords, visibility/indented methods, Python def/class),
//! builds an indent-based tree, and emits per-symbol chunks sharing a symbol
//! path like `CheckpointLoader > loadManifest()`.

use std::sync::LazyLock;

use regex::Regex;

const DEFAULT_MAX_CHARS: usize = 1200;
const DEFAULT_MIN_CHARS: usize = 200;

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct Chunk {
    pub ordinal: i64,
    pub heading_path: Option<String>,
    pub content: String,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub struct ChunkOptions {
    pub max_chars: usize,
    pub min_chars: usize,
}

impl Default for ChunkOptions {
    fn default() -> Self {
        Self {
            max_chars: DEFAULT_MAX_CHARS,
            min_chars: DEFAULT_MIN_CHARS,
        }
    }
}

#[derive(Debug, Clone, PartialEq, Eq)]
struct RawChunk {
    heading_path: Option<String>,
    content: String,
}

fn char_len(text: &str) -> usize {
    text.chars().count()
}

static ATX_HEADING_RE: LazyLock<Regex> =
    LazyLock::new(|| Regex::new(r"^(#{1,6})\s+(.+?)\s*#*\s*$").unwrap());
static BLANK_LINE_RE: LazyLock<Regex> = LazyLock::new(|| Regex::new(r"\n{2,}").unwrap());

// ---------------------------------------------------------------------------
// Generic size-bounded splitting
// ---------------------------------------------------------------------------

/// Split text into pieces of at most `max_chars` characters.
/// Tier 1: paragraphs (blank-line separated). Tier 2: lines. Tier 3: hard slices.
fn split_to_size(text: &str, max_chars: usize) -> Vec<String> {
    let trimmed = text.trim();
    if trimmed.is_empty() {
        return Vec::new();
    }
    split_tier(trimmed, max_chars)
}

fn split_tier(text: &str, max_chars: usize) -> Vec<String> {
    if char_len(text) <= max_chars {
        return vec![text.to_string()];
    }
    let parts: Vec<&str> = BLANK_LINE_RE
        .split(text)
        .map(str::trim)
        .filter(|part| !part.is_empty())
        .collect();
    if parts.len() > 1 {
        let pieces: Vec<String> = parts.iter().flat_map(|p| split_tier(p, max_chars)).collect();
        return pack_pieces(&pieces, "\n\n", max_chars);
    }
    let lines: Vec<&str> = text.split('\n').map(str::trim_end).collect();
    if lines.len() > 1 {
        let pieces: Vec<String> = lines.iter().flat_map(|l| split_line(l, max_chars)).collect();
        return pack_pieces(&pieces, "\n", max_chars);
    }
    split_line(text, max_chars)
}

fn split_line(line: &str, max_chars: usize) -> Vec<String> {
    let chars: Vec<char> = line.chars().collect();
    if chars.len() <= max_chars {
        return vec![line.to_string()];
    }
    chars
        .chunks(max_chars)
        .map(|slice| slice.iter().collect())
        .collect()
}

fn pack_pieces(pieces: &[String], separator: &str, max_chars: usize) -> Vec<String> {
    let mut out: Vec<String> = Vec::new();
    let mut buffer = String::new();
    for piece in pieces {
        if buffer.is_empty() {
            buffer = piece.clone();
        } else if char_len(&buffer) + separator.len() + char_len(piece) <= max_chars {
            buffer.push_str(separator);
            buffer.push_str(piece);
        } else {
            out.push(std::mem::take(&mut buffer));
            buffer = piece.clone();
        }
    }
    if !buffer.is_empty() {
        out.push(buffer);
    }
    out
}

// ---------------------------------------------------------------------------
// Markdown chunking
// ---------------------------------------------------------------------------

/// Split on ATX headings (`#`..`######` at line start), tracking the full
/// ancestor chain for the heading path.
fn split_markdown_sections(body: &str) -> Vec<RawChunk> {
    struct Accumulator<'a> {
        path: Option<Vec<String>>,
        lines: Vec<&'a str>,
    }

    fn flush(acc: &Accumulator<'_>, out: &mut Vec<RawChunk>) {
        let content = acc.lines.join("\n");
        let trimmed = content.trim();
        if trimmed.is_empty() {
            return;
        }
        out.push(RawChunk {
            heading_path: acc.path.as_ref().map(|p| p.join(" > ")),
            content: trimmed.to_string(),
        });
    }

    let mut sections: Vec<RawChunk> = Vec::new();
    let mut stack: Vec<(usize, String)> = Vec::new();
    let mut current = Accumulator {
        path: None,
        lines: Vec::new(),
    };

    for line in body.split('\n') {
        let Some(caps) = ATX_HEADING_RE.captures(line) else {
            current.lines.push(line);
            continue;
        };
        flush(&current, &mut sections);
        let level = caps[1].len();
        let title = caps[2].trim().to_string();
        while stack.last().is_some_and(|(l, _)| *l >= level) {
            stack.pop();
        }
        stack.push((level, title));
        current.path = Some(stack.iter().map(|(_, t)| t.clone()).collect());
        current.lines.clear();
    }
    flush(&current, &mut sections);
    sections
}

fn merge_small_chunks(chunks: &[RawChunk], min_chars: usize, max_chars: usize) -> Vec<RawChunk> {
    let mut merged: Vec<RawChunk> = Vec::new();
    for chunk in chunks {
        if let Some(previous) = merged.last_mut() {
            if char_len(&chunk.content) < min_chars
                && previous.heading_path == chunk.heading_path
                && char_len(&previous.content) + char_len(&chunk.content) <= max_chars * 3 / 2
            {
                previous.content.push_str("\n\n");
                previous.content.push_str(&chunk.content);
                continue;
            }
        }
        merged.push(RawChunk {
            heading_path: chunk.heading_path.clone(),
            content: chunk.content.clone(),
        });
    }
    merged
}

/// Chunk a markdown body into size-bounded chunks with heading paths.
pub fn chunk_markdown(body: &str, opts: ChunkOptions) -> Vec<Chunk> {
    let max_chars = opts.max_chars;
    let min_chars = opts.min_chars;
    let sections = split_markdown_sections(body);
    let merged = merge_small_chunks(&sections, min_chars, max_chars);
    let mut raw: Vec<RawChunk> = Vec::new();
    for chunk in merged {
        if char_len(&chunk.content) <= max_chars {
            raw.push(chunk);
        } else {
            for piece in split_to_size(&chunk.content, max_chars) {
                raw.push(RawChunk {
                    heading_path: chunk.heading_path.clone(),
                    content: piece,
                });
            }
        }
    }
    raw.into_iter()
        .enumerate()
        .map(|(index, chunk)| Chunk {
            ordinal: index as i64,
            heading_path: chunk.heading_path,
            content: chunk.content,
        })
        .collect()
}

// ---------------------------------------------------------------------------
// Code chunking
// ---------------------------------------------------------------------------

/// Map a filename extension to a language name; `None` for unknown or
/// missing extensions.
pub fn detect_language(filename: &str) -> Option<&'static str> {
    let base = filename.rsplit(['/', '\\']).next()?;
    let dot = base.rfind('.')?;
    if dot == 0 {
        return None;
    }
    Some(match base[dot + 1..].to_lowercase().as_str() {
        "ts" => "typescript",
        "tsx" => "tsx",
        "js" => "javascript",
        "jsx" => "jsx",
        "py" => "python",
        "go" => "go",
        "rs" => "rust",
        "java" => "java",
        "c" | "h" => "c",
        "cc" | "cpp" | "hpp" => "cpp",
        "cs" => "csharp",
        "rb" => "ruby",
        "php" => "php",
        "sh" => "shell",
        "sql" => "sql",
        "md" => "markdown",
        "json" => "json",
        "yaml" | "yml" => "yaml",
        "toml" => "toml",
        "html" => "html",
        "css" => "css",
        _ => return None,
    })
}

#[derive(Debug, Clone)]
struct DeclMatch {
    indent: usize,
    name: String,
    callable: bool,
    line_index: usize,
}

#[derive(Debug, Clone)]
struct DeclNode {
    decl: DeclMatch,
    span_end: usize,
    children: Vec<DeclNode>,
}

static KEYWORD_DECL_RE: LazyLock<Regex> = LazyLock::new(|| {
    Regex::new(
        r"^[ \t]*(?:export\s+)?(?:default\s+)?(?:declare\s+)?(?:abstract\s+)?(?:async\s+)?(?:pub\s+)?(class|function|struct|impl|fn|func|def)\s+([A-Za-z_$][A-Za-z0-9_$]*)",
    )
    .unwrap()
});
static MODIFIER_DECL_RE: LazyLock<Regex> = LazyLock::new(|| {
    Regex::new(
        r"^[ \t]*(?:public|private|protected)(?:\s+(?:static|readonly|async|override|final|abstract))*\s+([A-Za-z_$][A-Za-z0-9_$]*)\s*(?:<[^>\n]*>)?\(",
    )
    .unwrap()
});
static METHOD_DECL_RE: LazyLock<Regex> = LazyLock::new(|| {
    Regex::new(r"^[ \t]+([A-Za-z_$][A-Za-z0-9_$]*)\s*(?:<[^>\n]*>)?\([^;{}]*\)\s*(?::\s*[^{;=]+)?\{")
        .unwrap()
});
static PY_DECL_RE: LazyLock<Regex> = LazyLock::new(|| {
    Regex::new(r"^[ \t]*(?:async\s+)?(def|class)\s+([A-Za-z_][A-Za-z0-9_]*)").unwrap()
});
static COMMENT_LINE_RE: LazyLock<Regex> =
    LazyLock::new(|| Regex::new(r"^[ \t]*(//|/\*|\*|#)").unwrap());

const RESERVED_NAMES: [&str; 8] = [
    "if", "for", "while", "switch", "catch", "return", "new", "function",
];

const CALLABLE_KEYWORDS: [&str; 4] = ["function", "fn", "func", "def"];

fn indent_of(line: &str) -> usize {
    line.chars().take_while(|c| *c == ' ' || *c == '\t').count()
}

fn scan_decls(lines: &[&str], language_hint: Option<&str>) -> Vec<DeclMatch> {
    let python_only = language_hint == Some("python");
    let mut decls: Vec<DeclMatch> = Vec::new();
    for (i, line) in lines.iter().enumerate() {
        if COMMENT_LINE_RE.is_match(line) {
            continue;
        }

        if !python_only {
            if let Some(caps) = KEYWORD_DECL_RE.captures(line) {
                let keyword = &caps[1];
                decls.push(DeclMatch {
                    indent: indent_of(line),
                    name: caps[2].to_string(),
                    callable: CALLABLE_KEYWORDS.contains(&keyword),
                    line_index: i,
                });
                continue;
            }
        }

        if let Some(caps) = PY_DECL_RE.captures(line) {
            let keyword = &caps[1];
            decls.push(DeclMatch {
                indent: indent_of(line),
                name: caps[2].to_string(),
                callable: keyword == "def",
                line_index: i,
            });
            continue;
        }
        if python_only {
            continue;
        }

        if let Some(caps) = MODIFIER_DECL_RE.captures(line) {
            decls.push(DeclMatch {
                indent: indent_of(line),
                name: caps[1].to_string(),
                callable: true,
                line_index: i,
            });
            continue;
        }

        if let Some(caps) = METHOD_DECL_RE.captures(line) {
            let name = &caps[1];
            if !line.contains("=>") && !RESERVED_NAMES.contains(&name) {
                decls.push(DeclMatch {
                    indent: indent_of(line),
                    name: name.to_string(),
                    callable: true,
                    line_index: i,
                });
            }
        }
    }
    decls
}

fn build_tree(decls: &[DeclMatch], total_lines: usize) -> Vec<DeclNode> {
    let mut roots: Vec<DeclNode> = Vec::new();
    let mut stack: Vec<DeclNode> = Vec::new();

    fn pop_attach(stack: &mut Vec<DeclNode>, roots: &mut Vec<DeclNode>) {
        if let Some(popped) = stack.pop() {
            match stack.last_mut() {
                Some(parent) => parent.children.push(popped),
                None => roots.push(popped),
            }
        }
    }

    for (i, decl) in decls.iter().enumerate() {
        let mut span_end = total_lines;
        for later in &decls[i + 1..] {
            if later.indent <= decl.indent {
                span_end = later.line_index;
                break;
            }
        }
        let node = DeclNode {
            decl: decl.clone(),
            span_end,
            children: Vec::new(),
        };
        while matches!(stack.last(), Some(top) if top.decl.indent >= node.decl.indent) {
            pop_attach(&mut stack, &mut roots);
        }
        stack.push(node);
    }
    while !stack.is_empty() {
        pop_attach(&mut stack, &mut roots);
    }
    roots
}

fn display_name(decl: &DeclMatch) -> String {
    if decl.callable {
        format!("{}()", decl.name)
    } else {
        decl.name.clone()
    }
}

fn push_chunk(
    content: &str,
    heading_path: Option<&str>,
    out: &mut Vec<RawChunk>,
    max_chars: usize,
) {
    if char_len(content) <= max_chars {
        let trimmed = content.trim();
        if !trimmed.is_empty() {
            out.push(RawChunk {
                heading_path: heading_path.map(str::to_string),
                content: trimmed.to_string(),
            });
        }
        return;
    }
    for piece in split_to_size(content, max_chars) {
        out.push(RawChunk {
            heading_path: heading_path.map(str::to_string),
            content: piece,
        });
    }
}

fn emit_node(
    lines: &[&str],
    node: &DeclNode,
    ancestors: &[String],
    out: &mut Vec<RawChunk>,
    max_chars: usize,
) {
    let mut path: Vec<String> = ancestors.to_vec();
    path.push(display_name(&node.decl));
    let heading_path = path.join(" > ");

    if !node.children.is_empty() {
        let mut cursor = node.decl.line_index;
        for child in &node.children {
            let head = lines[cursor..child.decl.line_index].join("\n");
            push_chunk(&head, Some(&heading_path), out, max_chars);
            emit_node(lines, child, &path, out, max_chars);
            cursor = child.span_end;
        }
        let tail = lines[cursor..node.span_end].join("\n");
        push_chunk(&tail, Some(&heading_path), out, max_chars);
        return;
    }

    let body = lines[node.decl.line_index..node.span_end].join("\n");
    push_chunk(&body, Some(&heading_path), out, max_chars);
}

/// Chunk source code: declaration boundaries when symbols are detected
/// (coarse → fine via indent nesting), otherwise blank-line groups → lines →
/// hard slices. `language_hint` restricts Python to def/class patterns.
pub fn chunk_code(content: &str, language_hint: Option<&str>, opts: ChunkOptions) -> Vec<Chunk> {
    let max_chars = opts.max_chars;
    let lines: Vec<&str> = content.split('\n').collect();
    let decls = scan_decls(&lines, language_hint);
    let mut raw: Vec<RawChunk> = Vec::new();

    if decls.is_empty() {
        push_chunk(content, None, &mut raw, max_chars);
    } else {
        let roots = build_tree(&decls, lines.len());
        let first_line = decls[0].line_index;
        let preamble = lines[..first_line].join("\n");
        push_chunk(&preamble, None, &mut raw, max_chars);
        for root in &roots {
            emit_node(&lines, root, &[], &mut raw, max_chars);
        }
    }

    raw.into_iter()
        .enumerate()
        .map(|(index, chunk)| Chunk {
            ordinal: index as i64,
            heading_path: chunk.heading_path,
            content: chunk.content,
        })
        .collect()
}
